src/Control/Monad/Bayes/Population.hs

```diff
@@ -269,8 +273,12 @@ popAvg f p = do
 
 -- | Applies a transformation to the inner monad.
 hoist ::
-  Monad n =>
+  (Monad m, Monad n) =>
   (forall x. m x -> n x) ->
   Population m a ->
   Population n a
 hoist f = fromWeightedList . f . population
+
+-- | Flatten all layers of the free structure
+flatten :: Monad m => Population m a -> Population m a
```

Comment:
There is a design space here. Instead of making this an internal operation of `Population` which is invisible to the type checker, we could also have a second newtype:

```haskell
newtype FlatPopulation m a = FlatPopulation (m [a])
  deriving (Functor, Applicative) -- Note, no Monad
```
(Naming is up for grabs, we could also call the other thing `FreePopulation`.)
Then there is a function `Population m a -> FlatPopulation m a`. This has the advantage that we can specify in the different algorithms like `rmsmc`, `resampleGeneric` etc. when we need the structure flattened. That way we get a better grip on not making any mistakes with the order.